sdk/resource/container.go

```diff
@@ -0,0 +1,130 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource // import "go.opentelemetry.io/otel/sdk/resource"
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"io"
+	"os"
+	"strings"
+
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+)
+
+type containerIDProvider func() (string, error)
+
+var (
+	defaultContainerIDProvider containerIDProvider = getContainerIDFromCGroup
+)
+
+type containerIDDetector struct{}
+
+const cgroupPath = "/proc/self/cgroup"
```

Comment:
Given that this detector uses a specific technique to find this container's ID—this one assuming that one extraced from the cgroup file is the right choice—would you be amenable to naming this detector more specifically? Perhaps `cgroupContainerIDDetector`? That would clarify where this ID is coming from, and open us to other approaches to finding a container's ID.

I'm not sure whether it would warrant a corresponding change to the proposed `WithContainerID` function.